### PR TITLE
feat(trace): experimental traces for our tests

### PR DESCRIPTION
**Pull request by dgozman**
This introduces basic tracing enabled in our tests.

What is captured:
- network resources;
- snapshots at the start of most actions;
- snapshot after the test failure.

How this integrates with test runner:
- context fixture calls private method `context._initSnapshotter()` and uses `Tracer` to trace all events;
- all tests share a single `test-results/trace-storage` directory to store blobs;
- each test has its own trace file.

`npm run show-trace` opens a bare-minimum trace viewer that renders snapshots.

### CODE REVIEW

File: test/showTrace.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
+/**
```
Comment (pavelfeldman) [inline test/showTrace.js:1]: What does it do under the test folder?

File: test/playwright.fixtures.ts [hunk 2]
```diff
@@ -182,20 +182,29 @@ registerWorkerFixture('golden', async ({browserName}, test) => {
...
+registerFixture('page', async ({context, toImpl}, runTest, config, test) => {
   const page = await context.newPage();
   await runTest(page);
   if (test.error) {
     const relativePath = path.relative(config.testDir, test.file).replace(/\.spec\.[jt]s/, '');
     const sanitizedTitle = test.title.replace(/[^\w\d]+/g, '_');
     const assetPath = path.join(config.outputDir, relativePath, sanitizedTitle) + '-failed.png';
     await page.screenshot({ path: assetPath });
+    if (toImpl)
+      await toImpl(page)._captureSnapshot({ timeout: 5000, label: 'Test Failed' });
```
Comment (pavelfeldman) [inline test/playwright.fixtures.ts:207]: Let's make it available for real

File: test-runner/src/builtin.fixtures.ts [hunk 3]
```diff
@@ -57,6 +57,7 @@ const removeFolderAsync = promisify(rimraf);
 declare global {
   interface FixtureParameters {
     parallelIndex: number;
+    outputDir: string;
```
Comment (pavelfeldman) [inline test-runner/src/builtin.fixtures.ts:60]: outputDir is a part of config, it is not a fixture.

File: src/utils/traceTypes.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+  resources: {
+    url: string,
+    resourceId: string,
+    overrideSha1?: string,
+  }[],
+};
+
+export type PageSnapshot = {
+  viewportSize?: { width: number, height: number },
+  mainFrame: FrameSnapshot,
```
Comment (pavelfeldman) [inline src/utils/traceTypes.ts:70]: If this is a json structure, make first frame in frames a main frame?

File: src/utils/traceTypes.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+  shadowContent: { chunkId: string, html: string }[],
+
+  // <link href="URL">
+  // <img src="URL">
+  // <img srcset="URL">
+  // <... style="...url(URL)..." >
+  resources: {
+    url: string,
+    resourceId: string,
+    overrideSha1?: string,
```
Comment (pavelfeldman) [inline src/utils/traceTypes.ts:64]: What does it mean 'override sha1'?

File: src/server/trace/traceFile.ts [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+type AppendEventChain = { counter: number, promise: Promise<string> };
+const appendEventChains = new Map<string, AppendEventChain>();
+
+export class TraceFile {
+  private _traceStoragePromise: Promise<string>;
+  private _appendEventChain: AppendEventChain;
+
+  constructor(traceStorageDir: string, traceFile: string) {
+    this._traceStoragePromise = mkdirIfNeeded(path.join(traceStorageDir, 'sha1')).then(() => traceStorageDir);
+    const chain = appendEventChains.get(traceFile);
```
Comment (pavelfeldman) [inline src/server/trace/traceFile.ts:38]: This is completely unclear! What is happening here?

File: src/utils/traceTypes.ts [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+  }[],
+};
+
+export type PageSnapshot = {
```
Comment (pavelfeldman) [inline src/utils/traceTypes.ts:68]: I did not find a trace spec / type where `PageSnapshot` would be mentioned.

File: src/server/trace/traceController.ts [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+import { Page } from '../page';
+import * as network from '../network';
+import { helper, RegisteredListener } from '../helper';
+import { SavedResource, resourceReferences, captureSnapshot } from './snapshotter';
+import { Progress } from '../progress';
+import { ContextCreatedTraceEvent, ContextDestroyedTraceEvent, SnapshotTraceEvent, NetworkResponseTraceEvent } from '../../utils/traceTypes';
+import { TraceFile } from './traceFile';
+import { debugLogger } from '../../utils/debugLogger';
+
+export class TraceController {
```
Comment (pavelfeldman) [inline src/server/trace/traceController.ts:27]: Never understood the controller name - is this a recorder?

File: src/server/trace/traceController.ts [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+    await this.appendTraceEvent(snapshotEvent);
+  }
+
+  private _onPage(page: Page) {
+    this._eventListeners.push(helper.addEventListener(page, Page.Events.Response, (response: network.Response) => {
+      this._saveResource(response).catch(e => debugLogger.log('error', e));
+    }));
+  }
+
+  private async _saveResource(response: network.Response) {
```
Comment (pavelfeldman) [inline src/server/trace/traceController.ts:97]: This is getting close to har...

File: src/server/trace/snapshotter.ts [hunk 10]
```diff
@@ -0,0 +1,479 @@
...
+export type SavedResource = {
+  frameId: string,
+  url: string,
+  resourceId: string,
+  references: Promise<string[]>,
+};
+export type SavedResourceSet = () => Set<SavedResource>;
+export type ResourceSaver = (body: Buffer) => Promise<string>;  // Returns sha1.
+
+async function addResource(savedResourceSet: SavedResourceSet, resourceSaver: ResourceSaver, frameSnapshot: FrameSnapshot, frame: frames.Frame, url: string, override: string | undefined): Promise<void> {
```
Comment (pavelfeldman) [inline src/server/trace/snapshotter.ts:35]: 3 complex input types - can we get a little more oop-y here?

File: src/server/trace/snapshotter.ts [hunk 11]
```diff
@@ -0,0 +1,479 @@
...
+  }
+
+  return {
+    viewportSize,
+    mainFrame: mainFrame.snapshot,
+    frames: childFrames,
+  };
+}
+
+function takeSnapshot(guid: string, shadowAttribute: string, removeNoScript: boolean): any[] {
```
Comment (pavelfeldman) [inline src/server/trace/snapshotter.ts:191]: Looks like this runs in the page context - extract it into a separate file.

File: src/server/browserContext.ts [hunk 12]
```diff
@@ -73,6 +75,11 @@ export abstract class BrowserContext extends EventEmitter {
       new DebugController(this);
   }
 
+  // Used by test runner.
+  _traceTo(traceStorageDir: string, traceFilePath: string) {
```
Comment (pavelfeldman) [inline src/server/browserContext.ts:79]: Let's make it external to the context?

File: src/client/traceViewer.ts [hunk 13]
```diff
@@ -0,0 +1,245 @@
+/**
```
Comment (pavelfeldman) [inline src/client/traceViewer.ts:1]: Let's get ourselves a separate dumping ground folder for these.

**Review by pavelfeldman — APPROVED** [review]

File: test/playwright.fixtures.ts (hunk 2 above)
Comment (dgozman) [reply test/playwright.fixtures.ts]: I'll do that once we are ready for some kind of api :)

**Pull request merged**